### ISSUE TITLE
More control over ProgressBar formatting

### ISSUE THREAD

**Issue by jmg049** [pytorch/ignite#3118]
## 🚀 Feature - Improved ``contrib.handlers.tqdm_logger.ProgressBar`` formatting and alignment

More configurable arguments/bar formatting of the available ``ProgressBar`` class. Specifically to allow for the alignment of multiple bars (train loop, test/validation loop).

The current functionality of the ``ProgressBar`` class modifies the bar format to include the ``current_epoch``/``total_epochs`` when  ``max_epochs > 1``. This mostly works well for the train loop. However, format/alignment issues arise when the train loop is used with a validation loop that runs at the end of each epoch (on ``EpochCompleted``). The validation loop should only be run once. However, this causes the epoch information to be omitted, and consequently, any format options which attempt to align the bars are void (see the screenshot below; both progress bars are set to have the same formatting). The format given below may not format as intended, but I cannot test it since I cannot remove/add the epoch information. 

Ideally, control would be given over the if/how the epoch information is displayed when creating the ``ProgressBar``. Investigating the codebase, it seems that this stuff is handled internally in the ``_OutputHandler`` class. I would very much like to align the bars to start and finish in the same place as each other, purely for aesthetics. 

``BAR_FMT = "{desc:<15} [{n_fmt}/{total_fmt}]|{bar}| {rate_fmt}{postfix}"``

![image](https://github.com/pytorch/ignite/assets/121042936/31735c43-cb02-4d80-beb9-119e0ddcfad0)

In the above image, I would love control over aligning the actual bar sections of the progress bars.

**Comment by vfdev-5** [comment]
@jmg049 thanks for the suggestions! Would you like to send a draft PR with implementing that?

**Comment by jmg049** [comment]
> @jmg049 thanks for the suggestions! Would you like to send a draft PR with implementing that?

Hey, sure, it might take some time though unfortunately(if only contributing was my main work). I do already have a very hacky, massively imperfect prototype work-around for it.

But if I find some time I'll give it a shot. Definitely a niche, but fun and interesting problem.

**Comment by jmg049** [comment]
Yes it definitely looks better. No doubt about it. I'm still not a huge fan of always having the epoch info there and more specifically that it's tied to the specific Engine. 

I have something working (I think) that is quite nice, unfortunately the amount of time I have to work on it is small. Basically, I've added a custom field to the tqdm bar format, ``epochs``, which allows you to control the formatting of the epoch information. 

Then I've also added a parameter to the ``ProgressBar`` class named ``progress_with`` to allow for tying of something like and Evaluator progress bar to the Trainer progress bar. 

A (rough) example:
```python
trainer_pb = ProgressBar(...)
evaluator_pb = ProgressBar(...,  progress_with=trainer_pb)
```

Then the evaluator progress bar displays the same epoch information as the trainer.

**Comment by nithish-74** [comment]
Hi, I would like to work on this issue.

I am currently exploring the Ignite codebase while preparing a GSoC proposal related to improving training workflows. This issue looks like a great starting point, especially around improving usability and flexibility of ProgressBar.

My initial idea is to introduce configurable control over epoch display (e.g., a show_epoch flag), and possibly explore linking progress bars for better alignment between training and evaluation.

Please let me know if this direction sounds good or if there are specific expectations I should follow before starting.

Thanks!

**Comment by Aanushka001** [comment]
placeholder

**Comment by deodharaditi** [comment]
Hi, I'd like to work on this as part of a CodePath's AI301 program. I've read the thread and PR #3761 (and why it was closed), so I want to align on the approach before writing any code.

From the discussion, the real goal is letting a validation/eval ProgressBar align with the training bar - including consistent epoch info rather than just exposing more format strings. @jmg049's idea of a progress_with=<trainer_pbar> parameter (so a secondary bar inherits the trainer's epoch context) seems closer to what's wanted than a standalone show_epoch/epoch_format flag.

Proposed direction:

Add an opt-in way to control whether/how epoch info is rendered in ProgressBar (via _OutputHandler).
Add a progress_with= option to link a secondary bar to a primary one for aligned epoch display.
Cover it with tests in the existing tqdm_logger test module and run them locally before opening a PR.

Does this direction sound right, or would you prefer a narrower scope to start? Happy to adjust before I begin.

**Comment by vfdev-5** [comment]
I suggest to close this issue as there is no clear path to what we want to do here. Once we have something appropriate we can reopen this issue and implement the improvement. 
Currently, using `progress_with` arg path does look good to me due to very narrow and specific usage and potential errors.
